File: framework/languages/java/samples/java/GameQuest/Server/QuestMission/src/main/java/systems/zlink/samples/gamequest/server/questmission/domain/QuestDomain.java
package systems.zlink.samples.gamequest.server.questmission.domain;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

public final class QuestDomain {
    public QuestDecision apply(
        Messages.GameplayMsg event,
        List<Messages.QuestProgress> current,
        long nextVersion) {
        long now = Instant.now().toEpochMilli();
        List<Messages.StoredQuestEvent> stored = new ArrayList<>();
        List<Messages.QuestProgressNotify> progress = new ArrayList<>();
        List<Messages.QuestCompletedNotify> completed = new ArrayList<>();
        List<Messages.QuestProgress> updated = new ArrayList<>(current);

        if ("kill".equals(event.eventType()) && "wolf".equals(event.value())) {
            applyCounter(event, updated, stored, progress, completed, Messages.QuestIds.FirstHunt, 3, event.count(), now, nextVersion);
        } else if ("collect".equals(event.eventType()) && "healing-herb".equals(event.value())) {
            applyCounter(event, updated, stored, progress, completed, Messages.QuestIds.HerbGathering, 5, event.count(), now, nextVersion);
        } else if ("feature".equals(event.eventType()) && "auction".equals(event.value())) {
            completeOnce(event, updated, stored, completed, Messages.QuestIds.OpenAuction, 1, now, nextVersion);
        }

        return new QuestDecision(updated, stored, progress, completed);
    }

    private static void applyCounter(
        Messages.GameplayMsg event,
        List<Messages.QuestProgress> projection,
        List<Messages.StoredQuestEvent> stored,
        List<Messages.QuestProgressNotify> progressNotifications,
        List<Messages.QuestCompletedNotify> completedNotifications,
        String questId,
        int required,
        int delta,
        long now,
        long nextVersion) {
        Messages.QuestProgress previous = find(projection, questId);
        int current = Math.min(required, (previous == null ? 0 : previous.currentCount()) + delta);
        String status = current >= required
            ? Messages.QuestStatuses.RewardGranted
            : Messages.QuestStatuses.InProgress;
        Messages.QuestProgress next = new Messages.QuestProgress(
            event.playerId(),
            questId,
            status,
            current,
            required,
            event.eventId(),
            now);
        replace(projection, next);
        if (previous == null || previous.currentCount() != current) {
            stored.add(new Messages.StoredQuestEvent(
                event.eventId() + ":progress",
                event.eventId(),
                event.playerId(),
                questId,
                Messages.QuestProgressedEvent.class.getSimpleName(),
                next.currentCount(),
                next.requiredCount(),
                next.status(),
                nextVersion,
                now));
            if (event.publish()) {
                progressNotifications.add(new Messages.QuestProgressNotify(event.playerId(), null, next));
            }
        }
        if ((previous == null || !Messages.QuestStatuses.RewardGranted.equals(previous.status()))
            && Messages.QuestStatuses.RewardGranted.equals(status)) {
            stored.add(new Messages.StoredQuestEvent(
                event.eventId() + ":completed",
                event.eventId(),
                event.playerId(),
                questId,
                Messages.QuestCompletedEvent.class.getSimpleName(),
                next.currentCount(),
                next.requiredCount(),
                next.status(),
                nextVersion + 1,
                now));
            stored.add(new Messages.StoredQuestEvent(
                event.eventId() + ":reward",
                event.eventId(),
                event.playerId(),
                questId,
                Messages.QuestRewardGrantedEvent.class.getSimpleName(),
                next.currentCount(),
                next.requiredCount(),
                next.status(),
                nextVersion + 2,
                now));
            if (event.publish()) {
                completedNotifications.add(new Messages.QuestCompletedNotify(event.playerId(), null, next, true));
            }
        }
    }

    private static void completeOnce(
        Messages.GameplayMsg event,
        List<Messages.QuestProgress> projection,
        List<Messages.StoredQuestEvent> stored,
        List<Messages.QuestCompletedNotify> completedNotifications,
        String questId,
        int required,
        long now,
        long nextVersion) {
        Messages.QuestProgress previous = find(projection, questId);
        if (previous != null && Messages.QuestStatuses.RewardGranted.equals(previous.status())) {
            return;
        }
        Messages.QuestProgress next = new Messages.QuestProgress(
            event.playerId(),
            questId,
            Messages.QuestStatuses.RewardGranted,
            required,
            required,
            event.eventId(),
            now);
        replace(projection, next);
        stored.add(new Messages.StoredQuestEvent(
            event.eventId() + ":completed",
            event.eventId(),
            event.playerId(),
            questId,
            Messages.QuestCompletedEvent.class.getSimpleName(),
            next.currentCount(),
            next.requiredCount(),
            next.status(),
            nextVersion,
            now));
        stored.add(new Messages.StoredQuestEvent(
            event.eventId() + ":reward",
            event.eventId(),
            event.playerId(),
            questId,
            Messages.QuestRewardGrantedEvent.class.getSimpleName(),
            next.currentCount(),
            next.requiredCount(),
            next.status(),
            nextVersion + 1,
            now));
        if (event.publish()) {
            completedNotifications.add(new Messages.QuestCompletedNotify(event.playerId(), null, next, true));
        }
    }

    private static Messages.QuestProgress find(List<Messages.QuestProgress> projection, String questId) {
        return projection.stream()
            .filter(progress -> progress.questId().equals(questId))
            .findFirst()
            .orElse(null);
    }

    private static void replace(List<Messages.QuestProgress> projection, Messages.QuestProgress next) {
        projection.removeIf(progress -> progress.questId().equals(next.questId()));
        projection.add(next);
    }

    public record QuestDecision(
        List<Messages.QuestProgress> projection,
        List<Messages.StoredQuestEvent> storedEvents,
        List<Messages.QuestProgressNotify> progressNotifications,
        List<Messages.QuestCompletedNotify> completedNotifications) {
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/QuestMission/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/questmission/domain/QuestDomain.kt
package systems.zlink.samples.kotlin.gamequest.server.questmission.domain

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.json.JsonMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import systems.zlink.samples.kotlin.gamequest.server.configuration.QuestIds
import systems.zlink.samples.kotlin.gamequest.server.configuration.QuestStatuses
import systems.zlink.samples.kotlin.gamequest.shared.contracts.GameplayEventEnvelope
import systems.zlink.samples.kotlin.gamequest.shared.contracts.QuestCompletedEvent
import systems.zlink.samples.kotlin.gamequest.shared.contracts.QuestProgress
import systems.zlink.samples.kotlin.gamequest.shared.contracts.QuestProgressReconciledEvent
import systems.zlink.samples.kotlin.gamequest.shared.contracts.QuestProgressedEvent
import systems.zlink.samples.kotlin.gamequest.shared.contracts.QuestRewardGrantedEvent
import systems.zlink.samples.kotlin.gamequest.shared.contracts.StoredQuestEvent

/** Quest catalog and projection builder. Mirrors the .NET `QuestDomain`. */
object QuestDomain {
    private val json: ObjectMapper = JsonMapper.builder().build().registerKotlinModule()

    data class QuestDefinition(val questId: String, val eventType: String, val value: String, val required: Int)

    val CATALOG: List<QuestDefinition> = listOf(
        QuestDefinition(QuestIds.FirstHunt, "MonsterKilled", "*", 3),
        QuestDefinition(QuestIds.OpenAuction, "FeatureUnlocked", "auction", 1),
        QuestDefinition(QuestIds.HerbGathering, "ItemCollected", "healing-herb", 5),
        QuestDefinition(QuestIds.ClearTutorial, "MissionCompleted", "tutorial", 1),
        QuestDefinition(QuestIds.VisitRuins, "AreaEntered", "ruins", 1),
    )

    fun match(event: GameplayEventEnvelope): QuestDefinition? =
        CATALOG.firstOrNull {
            it.eventType == event.eventType && (it.value == "*" || it.value == event.value)
        }

    fun firstHunt(): QuestDefinition =
        CATALOG.first { it.questId == QuestIds.FirstHunt }

    fun apply(
        definition: QuestDefinition,
        current: QuestProgress?,
        event: GameplayEventEnvelope,
    ): QuestProgress {
        val currentCount = current?.currentCount ?: 0
        val nextCount = if (event.eventType == "SnapshotKillCount") {
            maxOf(currentCount, event.count)
        } else {
            minOf(definition.required, currentCount + event.count)
        }
        val complete = nextCount >= definition.required
        return QuestProgress(
            event.playerId,
            definition.questId,
            if (complete) QuestStatuses.RewardGranted else QuestStatuses.Active,
            nextCount,
            definition.required,
            event.eventId,
            System.currentTimeMillis(),
        )
    }

    fun toEvents(
        definition: QuestDefinition,
        before: QuestProgress?,
        after: QuestProgress,
        source: GameplayEventEnvelope,
    ): List<StoredQuestEvent> {
        if (before != null && source.eventId == before.lastEventId) {
            return emptyList()
        }
        if (before != null && before.currentCount == after.currentCount && before.status == after.status) {
            return emptyList()
        }

        val events = mutableListOf<StoredQuestEvent>()
        val kind = if (source.eventType == "SnapshotKillCount") {
            "QuestProgressReconciledEvent"
        } else {
            "QuestProgressedEvent"
        }
        events.add(create(kind, definition, before, after, source))

        val beforeStatus = before?.status
        if (beforeStatus != QuestStatuses.RewardGranted && after.status == QuestStatuses.RewardGranted) {
            events.add(create("QuestCompletedEvent", definition, before, after, source))
            events.add(create("QuestRewardGrantedEvent", definition, before, after, source))
        }
        return events
    }

    private fun create(
        eventType: String,
        definition: QuestDefinition,
        before: QuestProgress?,
        after: QuestProgress,
        source: GameplayEventEnvelope,
    ): StoredQuestEvent {
        val now = System.currentTimeMillis()
        val eventId = "${after.playerId}-${definition.questId}-${source.eventId}-$eventType"
        val payload: Any = when (eventType) {
            "QuestProgressedEvent" -> QuestProgressedEvent(
                eventId,
                after.playerId,
                definition.questId,
                maxOf(0, after.currentCount - (before?.currentCount ?: 0)),
                after.currentCount,
                after.requiredCount,
                source.eventId,
            )
            "QuestCompletedEvent" -> QuestCompletedEvent(
                eventId, after.playerId, definition.questId, source.eventId, now,
            )
            "QuestRewardGrantedEvent" -> QuestRewardGrantedEvent(
                eventId, after.playerId, definition.questId, source.eventId,
                "reward-${definition.questId}", now,
            )
            "QuestProgressReconciledEvent" -> QuestProgressReconciledEvent(
                eventId, after.playerId, definition.questId, after.currentCount, "GameplaySnapshot", now,
            )
            else -> throw IllegalArgumentException("Unknown quest event type: $eventType")
        }
        return StoredQuestEvent(
            eventId,
            source.eventId,
            after.playerId,
            definition.questId,
            eventType,
            json.writeValueAsBytes(payload),
            0,
            now,
        )
    }
}
